File: 백준/Silver/1972. 놀라운 문자열/놀라운 문자열.py
import sys
input = sys.stdin.read().strip().splitlines()

def check(depth, word):
    start_set = set()
    for i in range(len(word) - depth - 1):
        start = word[i] + word[i + depth + 1]
        if start in start_set:
            return False
        start_set.add(start)
    return True

index = 0
answer = []
while True:
    word = input[index]
    index += 1
    if word == "*":
        break

    N = len(word)
    is_surprising = True
    for i in range(N - 1):
        if not check(i, word):
            is_surprising = False
            break

    if is_surprising:
        answer.append(f"{word} is surprising.")
    else:
        answer.append(f"{word} is NOT surprising.")

print("\n".join(answer))

File: 백준/Silver/1972. 놀라운 문자열/놀라운 문자열.js
const input = require("fs").readFileSync("/dev/stdin").toString().trim().split("\n")

const answer = []

const check = (depth, word, isAlreadyIn) => {
  let start = ""
  isAlreadyIn.push(start)
  const set = new Set()

  for (let i = 0; i < word.length - depth - 1; i++) {
    start = word[i] + word[i + depth + 1]
    if (set.has(start)) {
      return false
    }
    set.add(start)
  }
  return true
}

let index = 0
while (true) {
  const word = input[index++]
  if (word === "*") break

  const N = word.length
  let flag = true
  for (let i = 0; i <= N - 2; i++) {
    if (!check(i, word, [])) {
      flag = false
      break
    }
  }

  if (flag === false) {
    answer.push(`${word} is NOT surprising.`)
  } else {
    answer.push(`${word} is surprising.`)
  }
}
console.log(answer.join("\n"))
